### PR TITLE
Fix routing issues when loading a `Router` via a dynamic library

### PR DESCRIPTION
**Pull request by davidpdrsn**
Fixes https://github.com/tokio-rs/axum/issues/1803

I also tested it with the repro script and does indeed fix the problem.

### CODE REVIEW

**Comment by davidpdrsn** [comment]
@Curid do you wanna test this to double check that it works?

File: axum/src/routing/mod.rs [hunk 1]
```diff
@@ -47,24 +47,12 @@ pub use self::method_routing::{
...
-        Self(id)
-    }
-}
-
 /// The router type for composing handlers and services.
 pub struct Router<S = (), B = Body> {
     routes: HashMap<RouteId, Endpoint<S, B>>,
     node: Arc<Node>,
     fallback: Fallback<S, B>,
+    prev_route_id: Option<RouteId>,
```
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply axum/src/routing/mod.rs:55]: Oh yeah we could just default to `RouteId(0)` 👍

File: axum/src/routing/mod.rs [hunk 2]
```diff
@@ -506,6 +500,16 @@ where
...
     }
+
+    fn next_route_id(&mut self) -> RouteId {
+        let next_id = self
+            .prev_route_id
+            .0
+            .checked_add(1)
+            .expect("Over `u32::MAX` routes created. If you need this, please file an issue.");
+        self.prev_route_id = RouteId(next_id);
+        self.prev_route_id
```
Comment (jplatte) [inline axum/src/routing/mod.rs:511]: I still we should name this `next_route_id` and return `RouteId(0)` as the first one as before, not `RouteId(1)`. But that's very nitpicky at this point.

**Review by jplatte — APPROVED** [review]

File: axum/src/routing/mod.rs (hunk 2 above)
Comment (davidpdrsn) [reply axum/src/routing/mod.rs:511]: Hehe I think I'll just leave it like this.

**Comment by davidpdrsn** [comment]
I'll just merge this. @Curid let us know if it doesn't work for you :)

**Pull request merged**